### PR TITLE
Implement function that allows parsing executable script into variables

### PR DESCRIPTION
**Pull request by Benzbeeb**

### CODE REVIEW

File: scan/__tests__/Paser_test.re [hunk 1]
```diff
@@ -0,0 +1,58 @@
+open Jest;
+open Parser;
+
+open Expect;
+
+describe("Expect Parser to work correctly", () => {
+  test("should be able to create Hash from hex", () =>
```
Comment (taobun) [inline scan/__tests__/Paser_test.re]: What?

File: scan/__tests__/Paser_test.re [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    expect("benz=$1" |> checker) |> toEqual(true)
+  );
+
+  test("should be able to create Hash from hex", () =>
```
Comment (taobun) [inline scan/__tests__/Paser_test.re]: ditto

**Review by taobun — CHANGES_REQUESTED** [review]

File: scan/src/utils/Parser.re [hunk 3]
```diff
@@ -0,0 +1,18 @@
+let checker = (str: string) => {
+  let len = String.length(str);
+  len >= 3 && str.[len - 3] == '=' && str.[len - 2] == '$' && str.[len - 1] == '1';
```
Comment (sorawit) [inline scan/src/utils/Parser.re]: not just 1, any number should do

**Review by sorawit — CHANGES_REQUESTED** [review]

**Comment by sorawit** [comment]
Let's call the file "ExecutableParser"

File: scan/__tests__/ExecutableParser_test.re [hunk 4]
```diff
@@ -0,0 +1,58 @@
+open Jest;
+open ExecutableParser;
+
+open Expect;
+
+describe("Expect Parser to work correctly", () => {
+  test("test check regex", () =>
+    expect("benz=$1" |> checker) |> toEqual(true)
+  );
+
```
Comment (taobun) [inline scan/__tests__/ExecutableParser_test.re:22]: Test not match regex too!

File: scan/__tests__/ExecutableParser_test.re [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+  test("test getVariables", () =>
```
Comment (taobun) [inline scan/__tests__/ExecutableParser_test.re]: test get 1 variable

File: scan/__tests__/ExecutableParser_test.re [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    expect({f|"#!/bin/bash
+
+calldata=$1
+
+echo $calldata
+"|f} |> getVariables)
+    |> toEqual(["calldata"])
+  );
+
+  test("test getVariables", () =>
```
Comment (taobun) [inline scan/__tests__/ExecutableParser_test.re]: Test many variables

File: scan/__tests__/ExecutableParser_test.re [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+  test("test parseExecutableScript", () =>
+    expect(
+      "IyEvYmluL3NoCgpzeW1ib2w9JDEKCiMgQ3J5cHRvY3VycmVuY3kgcHJpY2UgZW5kcG9pbnQ6IGh0dHBzOi8vd3d3LmNvaW5nZWNrby5jb20vYXBpL2RvY3VtZW50YXRpb25zL3YzCnVybD0iaHR0cHM6Ly9hcGkuY29pbmdlY2tvLmNvbS9hcGkvdjMvc2ltcGxlL3ByaWNlP2lkcz0kc3ltYm9sJnZzX2N1cnJlbmNpZXM9dXNkIgoKIyBQZXJmb3JtcyBkYXRhIGZldGNoaW5nIGFuZCBwYXJzZXMgdGhlIHJlc3VsdApjdXJsIC1zIC1YIEdFVCAkdXJsIC1IICJhY2NlcHQ6IGFwcGxpY2F0aW9uL2pzb24iIHwganEgLWVyICIuW1wiJHN5bWJvbFwiXS51c2QiCg"
+      |> JsBuffer.fromBase64
+      |> parseExecutableScript,
+    )
+    |> toEqual(["symbol"])
+  );
+});
+// IyEvYmluL3NoCgpzeW1ib2w9JDEKCiMgQ3J5cHRvY3VycmVuY3kgcHJpY2UgZW5kcG9pbnQ6IGh0dHBzOi8vd3d3LmNvaW5nZWNrby5jb20vYXBpL2RvY3VtZW50YXRpb25zL3YzCnVybD0iaHR0cHM6Ly9hcGkuY29pbmdlY2tvLmNvbS9hcGkvdjMvc2ltcGxlL3ByaWNlP2lkcz0kc3ltYm9sJnZzX2N1cnJlbmNpZXM9dXNkIgoKIyBQZXJmb3JtcyBkYXRhIGZldGNoaW5nIGFuZCBwYXJzZXMgdGhlIHJlc3VsdApjdXJsIC1zIC1YIEdFVCAkdXJsIC1IICJhY2NlcHQ6IGFwcGxpY2F0aW9uL2pzb24iIHwganEgLWVyICIuW1wiJHN5bWJvbFwiXS51c2QiCg
```
Comment (taobun) [inline scan/__tests__/ExecutableParser_test.re]: Delete comment

**Review by taobun — CHANGES_REQUESTED** [review]

File: scan/src/utils/ExecutableParser.re [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+
+let trim = (str: string) => {
+  str |> String.split_on_char('=') |> Belt_List.get(_, 0) |> Belt_Option.getExn;
+};
+
+let getVariables = (str: string) => {
+  String.split_on_char('\n', str) |> List.filter(checker) |> List.map(trim);
+};
+
+let parseExecutableScript = (buff: JsBuffer.t) => {
```
Comment (sorawit) [inline scan/src/utils/ExecutableParser.re:48]: @taobun @evilpeach. What would be the ideal return type for this function for the UI to use?

Comment (sorawit) [reply scan/src/utils/ExecutableParser.re:48]: I think It also returns `["A", "B"]` if someone writes code
```
#/bin/sh

B=$2
A=$1

...
```

**Review by sorawit — CHANGES_REQUESTED** [review]

File: scan/src/utils/ExecutableParser.re [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+};
+
+let splitToPair = s => {
+  let tmp = s |> String.split_on_char('=');
+  let s0 = getElementInList(tmp, 0);
+  let s1 =
+    getElementInList(tmp, 1)
+    |> String.split_on_char('$')
+    |> getElementInList(_, 1)
+    |> int_of_string;
```
Comment (sorawit) [inline scan/src/utils/ExecutableParser.re]: This will crash if number is too big
```
symbol=$123123091823901283019381
```

**Review by sorawit — CHANGES_REQUESTED** [review]

File: scan/src/utils/ExecutableParser.re (hunk 9 above)
Comment (Benzbeeb) [reply scan/src/utils/ExecutableParser.re]: `If the input exceeds max_int it is converted to the signed integer min_int + input - max_int - 1. `

File: scan/__tests__/ExecutableParser_test.re [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+
+echo $calldata
+"|f}
+      |> getVariables
+      |> Belt_Option.getExn,
+    )
+    |> toEqual(["calldata", "symbol"])
+  );
+
+  test("test get many variables", () =>
```
Comment (taobun) [inline scan/__tests__/ExecutableParser_test.re]: test invalid order variables

**Review by taobun — CHANGES_REQUESTED** [review]

File: scan/src/utils/ExecutableParser.re [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+let splitToPair = s => {
+  let tmp = s |> String.split_on_char('=');
+  let s0 = getElementInList(tmp, 0);
+  let s1 =
+    getElementInList(tmp, 1)
+    |> String.split_on_char('$')
+    |> getElementInList(_, 1)
+
+  switch (s1 |> int_of_string) {
+    // the number is to big then return -1
```
Comment (sorawit) [inline scan/src/utils/ExecutableParser.re]: too big

File: scan/src/utils/ExecutableParser.re [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+    | exception (Failure(_)) => (s0,-1)
```
Comment (sorawit) [inline scan/src/utils/ExecutableParser.re]: please format the code properly

**Review by sorawit — CHANGES_REQUESTED** [review]

**Review by taobun — APPROVED** [review]

**Review by sorawit — APPROVED** [review]

**Pull request merged**